src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypingStyles
+{
+    internal abstract partial class CSharpTypingStyleDiagnosticAnalyzerBase : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        [Flags]
+        internal enum TypingStyles
+        {
+            None = 0,
+            VarForIntrinsic = 1 << 0,
+            VarWhereApparent = 1 << 1,
+            VarWherePossible = 1 << 2,
+        }
+
+        private readonly string _diagnosticId;
+        private readonly LocalizableString _title;
+        private readonly LocalizableString _message;
+        private readonly Lazy<DiagnosticDescriptor> _noneDiagnosticDescriptor;
+        private readonly Lazy<DiagnosticDescriptor> _infoDiagnosticDescriptor;
+        private readonly Lazy<DiagnosticDescriptor> _warningDiagnosticDescriptor;
+        private readonly Lazy<DiagnosticDescriptor> _errorDiagnosticDescriptor;
+        private readonly Dictionary<DiagnosticSeverity, Lazy<DiagnosticDescriptor>> _severityToDescriptorMap;
+
+        public CSharpTypingStyleDiagnosticAnalyzerBase(string diagnosticId, LocalizableString title, LocalizableString message)
+        {
+            _diagnosticId = diagnosticId;
+            _title = title;
+            _message = message;
+            _noneDiagnosticDescriptor = new Lazy<DiagnosticDescriptor>(() => CreateDiagnosticDescriptor(DiagnosticSeverity.Hidden));
```

Comment:
This lazyness is not saving us anything.  We now have more allocatoins because of hte lazy and hte delegate.  just new up the diagnostic descriptor.  Having 4 extra objects for them is not a problem.